File: sort/quick/quick_sort.go
package main

import "fmt"

//quick_sort
func quick_sort(arr []int, left, right int) {
	if left < right {
		pivot := partition(arr, left, right)
		quick_sort(arr, left, pivot-1)
		quick_sort(arr, pivot+1, right)
	}
}

func partition(arr []int, left, right int) int {
	pivot := arr[right]
	i := left - 1
	for j := left; j < right; j++ {
		if arr[j] <= pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
	arr[i+1], arr[right] = arr[right], arr[i+1]
	return i + 1
}

func main() {
	array := []int{5, 7, 9, 0, 3, 1, 6, 2, 4, 8}
	quick_sort(array, 0, len(array)-1)
	for _, v := range array {
		fmt.Printf("%d ", v)
	}
}

File: basic_argorithm/Sort/quick_sort.py
arr = [9, 7, 5, 3, 1, 2, 4, 8, 6, 10]

def quick_sort(arr, start, end):
    if start >= end:
        return

    pivot = arr[start]
    i = start + 1
    j = end

    while i <= j:
        while i <= end and arr[i] <= pivot:
            i += 1
        while j > start and arr[j] >= pivot:
            j -= 1

        if i > j:
            arr[start], arr[j] = arr[j], arr[start]
        else:
            arr[i], arr[j] = arr[j], arr[i]

    quick_sort(arr, start, j - 1)
    quick_sort(arr, j + 1, end)

    return arr

print(quick_sort(arr, 0, len(arr) - 1))


def quick_sort2(arr):
    if len(arr) <= 1:
        return arr
    
    pivot = arr[0]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]

    return quick_sort2(left) + middle + quick_sort2(right)

print(quick_sort2(arr))